content/blog/2025-05-15-multiprocess.mdx

```diff
@@ -0,0 +1,144 @@
+In `vfork()`, the child executes at the same address space as the parent until the child calls `execve()` or terminates.
+For the sake of consistency, the parent blocks until one of these events occurs.
+Moreover, as any change by the child will be eventually visible to the parent, the child is only allowed a limited set of actions:
+(1) to update the parent's return value with the child's pid, (2) call `execve()`, and (3) call `_exit()`.
+The result of any other behavior is undefined.
+
+Linux introduced the non-standard  [clone(2)](https://man7.org/linux/man-pages/man2/clone.2.html) syscall, that included the `CLONE_VFORK` (suspend parent) and `CLONE_VM` (calling process and child shares the same memory space) flags.
+With these flags, vfork() could be easily implemented by calling `clone()` with flags set to `CLONE_VM | CLONE_VFORK | SIGCHLD`.
+
+[posix_spawn(3)](https://www.man7.org/linux/man-pages/man3/posix_spawn.3.html) is a libc function that performs the steps required to spawn a new process along with other housekeeping tasks in a safe and performant way. Because of that, it is the recommended way to spawn a new process.
+Under the hood, the various implementation of libc use `fork()`, `vfork()`, or `clone()`.
+musl libc [implements posix_spawn()](https://github.com/esmil/musl/blob/master/src/process/posix_spawn.c) using clone().
+glibc has some interesting history: Until 2.23 it used to provide the `POSIX_SPAWN_USEVFORK` flag to use `vfork()` over `fork()`.
+Starting from glibc 2.24, it switched from `vfork()` to `clone()`, and in recent versions `POSIX_SPAWN_USEVFORK` has no effect.
+Android's Bionic libc [still honors](https://android.googlesource.com/platform/bionic/+/master/libc/bionic/spawn.cpp#195) `POSIX_SPAWN_USEVFORK`.
+
+POSIX.1-2001 marked `vfork()` as obsolete in favor of `fork()` with COW and `posix_spawn()`, until it was finally removed in POSIX.1-2008.
+Nevertheless, Linux [continues to support](https://elixir.bootlin.com/linux/v6.11/source/kernel/fork.c#L2879) `vfork()` to this day, and so do the [various](https://man.freebsd.org/cgi/man.cgi?query=vfork) [flavors](https://www.netbsd.org/docs/kernel/vfork.html) of [BSD](https://man.openbsd.org/vfork.2), as `vfork()` remains relevant in systems without an MMU, performance-critical applications, and memory constrained systems.
+
+# The implementation on Unikraft
+
+Given the challenges of `fork()` and Unikraft’s single address space nature, we focus on supporting `posix_spawn`-like workflows, with a conscious decision against any full-fledged `fork()` support. We believe this strikes the right balance between compatibility and architectural purity.
+
+To enable this, [libposix-process](https://github.com/unikraft/unikraft/blob/staging/lib/posix-process) was carefully restructured so that multiprocess capabilities are not intrusive and Unikraft's modular nature is preserved.
```

Comment:
```suggestion
To enable this, [`libposix-process`](https://github.com/unikraft/unikraft/blob/staging/lib/posix-process) was carefully restructured so that multiprocess capabilities are not intrusive and Unikraft's modular nature is preserved.
```